src/analyz/mod.rs

```diff
@@ -95,37 +95,85 @@ fn vtable_descriptor_for_cast<'tcx>(
-    };
-    let new_pointee = match *new_ty.kind() {
-        ty::TyKind::Ref(_, ty, _) => ty,
-        ty::TyKind::RawPtr(ty, _) => ty,
-        _ => return None,
-    };
+    fn vtable_descriptor_for_ref_or_ptr_cast<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        old_pointee: ty::Ty<'tcx>,
+        new_pointee: ty::Ty<'tcx>,
+    ) -> Option<ty::PolyTraitRef<'tcx>> {
+        if !tcx.is_sized_raw(ty::TypingEnv::fully_monomorphized().as_query_input(old_pointee)) {
+            // We produce a vtable only for sized -> TyDynamic casts.
+            return None;
+        }
 
-    if !tcx.is_sized_raw(ty::TypingEnv::fully_monomorphized().as_query_input(old_pointee)) {
-        // We produce a vtable only for sized -> TyDynamic casts.
-        return None;
+        // Relevant code: rustc_codegen_ssa::meth::get_vtable
+        let trait_ref = match *new_pointee.kind() {
+            ty::TyKind::Dynamic(ref preds, _, _) =>
+                preds.principal().map(|pred| pred.with_self_ty(tcx, old_pointee)),
+            _ => return None,
+        };
+        let trait_ref: ty::PolyTraitRef = match trait_ref {
+            Some(x) => x,
+            // If there's no trait ref, it means the `Dynamic` predicates contain only auto traits.
+            // The vtable for this trait object is empty.
+            // TODO: handle this better (currently the output omits the "vtable" field)
+            None => return None,
+        };
+        Some(trait_ref)
     }
 
-    // Relevant code: rustc_codegen_ssa::meth::get_vtable
-    let trait_ref = match *new_pointee.kind() {
-        ty::TyKind::Dynamic(ref preds, _, _) =>
-            preds.principal().map(|pred| pred.with_self_ty(tcx, old_pointee)),
-        _ => return None,
-    };
-    let trait_ref: ty::PolyTraitRef = match trait_ref {
-        Some(x) => x,
-        // If there's no trait ref, it means the `Dynamic` predicates contain only auto traits.
-        // The vtable for this trait object is empty.
-        // TODO: handle this better (currently the output omits the "vtable" field)
-        None => return None,
-    };
-    Some(trait_ref)
+    fn instantiate_field_ty<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        field_def: &ty::FieldDef,
+        args: ty::GenericArgsRef<'tcx>
+    ) -> ty::Ty<'tcx> {
+        let field_unsubst_ty = tcx.type_of(field_def.did);
+        tcx.instantiate_and_normalize_erasing_regions(
+            args, ty::TypingEnv::fully_monomorphized(), field_unsubst_ty
+        )
+    }
+
+    match (*old_ty.kind(), *new_ty.kind()) {
+        // The two base cases: we are casting between two references or two raw
+        // pointers.
+        (ty::TyKind::Ref(_, old_pointee, _), ty::TyKind::Ref(_, new_pointee, _)) =>
+            vtable_descriptor_for_ref_or_ptr_cast(tcx, old_pointee, new_pointee),
+        (ty::TyKind::RawPtr(old_pointee, _), ty::TyKind::RawPtr(new_pointee, _)) =>
+            vtable_descriptor_for_ref_or_ptr_cast(tcx, old_pointee, new_pointee),
```

Comment:
```suggestion
        (ty::TyKind::Ref(_, old_pointee, _), ty::TyKind::Ref(_, new_pointee, _)) |
        (ty::TyKind::RawPtr(old_pointee, _), ty::TyKind::RawPtr(new_pointee, _)) =>
            vtable_descriptor_for_ref_or_ptr_cast(tcx, old_pointee, new_pointee),
```
And then you can move the code for `vtable_descriptor_for_ref_or_ptr_cast` into this case of the match